### PR TITLE
Add AssemblyExtensions.GetApplyUpdateCapabilities method

### LINKED ISSUE #50111: internal runtime API to return hot reload capabilities
MonoVM and CoreCLR will both support hot reload in .NET 6.  However the actual set of supported (non-"rude") edits are different between the runtimes.  In future releases the capabilities of both runtimes will evolve.  Part of https://github.com/dotnet/runtime/issues/44806 and https://github.com/dotnet/runtime/issues/45629

Related roslyn issue: https://github.com/dotnet/roslyn/issues/49010

In a hot reload editing session, the runtime and the Roslyn compiler (mediated by a communication mechanism such as `dotnet watch` or the Visual Studio debugger) will need to agree on the supported edits.

One approach would be to build into Roslyn the knowledge of the capabilities of each runtime based on the reported runtime version.  However in this case it would be difficult to mix compiler and runtime versions.  Also if the runtime wants to disable an edit in a servicing release, it would require closely coupling servicing releases of (potentially several) Roslyn versions and the runtime.

Another approach is to expose a new API implemented by the runtime that returns a list of the supported capabilities:

```csharp
 namespace System.Reflection.Metadata
{
    partial class AssemblyExtensions
    {
        /// Returns a description of the supported System.Reflection.Metadata.AssemblyExtensions.ApplyUpdate edits.
        /// Returns string.Empty if applying updates is not supported by this runtime.
        internal static string GetApplyUpdateCapabilities ();
    }
}
```

The specific format of the return value would be implementation defined and in practice it would be a contract between the runtimes and Roslyn.

### Example of capabilities description 

(This is for illustration only and not part of the API definition)

For example the string might be a space separated sequence of tokens from the following set:  `Baseline`, `RuntimeEdits`, `AddDefinitionToExistingType`, `NewTypeDefinition`. The proposal is to represent .NET 5 CoreCLR capabilities as `Baseline` plus everything that .NET 5 CoreCLR can do *minus* what .NET 6 MonoVM can do.  That is, .NET 5 MonoVM would be represented by the empty set of capabilities, .NET 5 CoreCLR as `Baseline` plus everything that CoreCLR can do that .NET6 Mono cannot.  In future releases, as MonoVM catches up it will start returning `AddDefinitionToExistingType` (for example).  In future releases when one or both runtimes add a new capability, it will be added to the enumeration.  In summary:

* .NET 5 CoreCLR, .NET Framework: `"Baseline AddDefinitionToExistingType NewTypeDefinition "` (and ... TBD based on Mono capabilities )
* .NET 6 CoreCLR: .NET 5 + `"RuntimeEdits"`
* .NET 5 MonoVM: `""`
* .NET 6 MonoVM: `"Baseline RuntimeEdits"`

### PR DESCRIPTION
**Pull request by lambdageek**
Implement a private API that returns the hot reload capabilities of the current runtime.

The corresponding Roslyn API work is https://github.com/dotnet/roslyn/pull/52566

Resolves #50111

### CODE REVIEW

**Review by mikem8361 — APPROVED** [review]

File: src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs [hunk 1]
```diff
@@ -39,5 +39,18 @@ public static void ApplyUpdateInvalidParameters()
...
+        public void GetApplyUpdateCapabilitiesIsCallable()
+        {
+            var ty = typeof(System.Reflection.Metadata.AssemblyExtensions);
+            var mi = ty.GetMethod("GetApplyUpdateCapabilities", BindingFlags.NonPublic | BindingFlags.Static, Array.Empty<Type>());
+
+            Assert.NotNull(mi);
+
+            var result = mi.Invoke(null, null);
+
+            Assert.NotNull(result);
```
Comment (marek-safar) [inline src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:53]: It's bit odd to check for notnull if the return value is not declared as nullable

Comment (jkotas) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:53]: Nullable annotations are just a suggestions. They do not guarantee any behavior. We have similar checks in many other tests.

Comment (jkotas) [inline src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:53]: Verify that the method returned a string?

**Review by jkotas — APPROVED** [review]
LGTM

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml [hunk 2]
```diff
@@ -62,4 +62,13 @@
       <method signature="System.Void Initialize()" />
     </type>
   </assembly>
+
+  <!-- methods used by hot reload.
+       TODO: once there's a feature flag, add it to the linker descriptor
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:67]: Can we use the `Debugger.IsSupported` feature switch for now? That's what we did for `MetadataUpdateHandlerAttribute`.

https://github.com/dotnet/runtime/blob/636988a0727d01479f9af94ef1c56bb7e79ecb24/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml#L33-L36

I can put a PR to fix this.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:67]: Yes, that makes sense to me.

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:67]: I opened https://github.com/dotnet/runtime/pull/51994

**Pull request merged**